smolyay/surrogate.py

```diff
@@ -1,546 +1,828 @@
+
+    Depending on the dimensionality (number of independent variables),
+    sampling method, and combination of basis functions, a surrogate
+    model can be generated that approximates a set of data.
+    ``domain`` is the domain of the function to be approximated.
+    ``basis_sets`` is a list of :class:BasisFunctionSets that describe
+    the basis functions used to construct terms in the surrogate function,
+    with the :class:BasisFunctionSet at index i in the list being the set of
+    basis functions used for the ith dimensional variable.
+    ``regularization`` is the method of regularization used in the event that the
+    number of terms does not match the number of sample points. Ridge
+    regularization and lasso regularization are available, as well as finding
+    the solution of least squares. If the basis functions return complex
+    values, then solving for least squares will be used instead of ridge
+    or lasso.
+
+    :attr:`index_combinations` describes the combination of basis function
+    used to construct the terms of the surrogate, where each row is a term
+    represented by a list of size ``num_dimensions`` that give the index of
+    the basis functions from the :class:BasisFunctionSets in `basis_sets`
+    that make up a given term.
+    :attr:`coefficients` is the list of coefficients for each term in the
+    surrogate equation. These cefficients are determined in :meth:`fit`.
+    :meth:`fit` computes the coefficients based on a set of data
+    (function values at transformed grid points).
+    Once the surrogate is constructed, one can evaluate the surrogate
+    through :meth:`predict(x)` and its gradient can be evaluated
+    through :meth:`gradient(x)`.
+    :meth:`_create_terms`, is an abstract method for generating the terms
+    described in :attr:`index_combinations`.
+
+    Parameters
+    ----------
+    domain: numpy.ndarray
+        the domain of the surrgoate
+
+    basis_sets: list of :class:BasisFunctionSet
+        the set of basis functions used to combine terms
+
+    regularization : [:class:L1Regularization, :class:L2Regularization, None], default None
+        the regularization method for determining the coefficients
+    """
+
+    def __init__(self, domain, basis_sets, regularization=None):
+        super().__init__(domain)
+        self._basis_sets = None
+        self._regularization = None
+        self._index_combinations = None
+        self._coefficients = None
+        self._integration_constant = 0
+        self._terms_constructed_cache = False
+
+        self._basis_sets = basis_sets
+        self.regularization = regularization
+
+    @property
+    def regularization(self):
+        """:class:RegularizationHelper: constant of the L2 and L1 term"""
+        return self._regularization
+
+    @regularization.setter
+    def regularization(self, value):
+        if not value is None:
+            if not isinstance(value, RegularizationHelper):
```

Comment:
This is the pattern you can follow for ifs like this. They get evaluated left to right, and later statements are not evaluated if earlier ones fail.
```suggestion
        if value is not None and not isinstance(value, RegularizationHelper):
```